File: api/app/modules/jdmatch/agents/extract_jd.py
import asyncio
import json

from browser_use_sdk import BrowserUse

from app.core.config import settings
from app.core.logging.logger import get_logger
from app.integrations.llm.gemini import GeminiModel

logger = get_logger("extract_jd.agent")

_static_prompt = """
Strictly follow this json format output:

Success:
{
  "message": "SUCCESS",
  "data": "<Extracted Content Of the JD>"
}

Error:
{
  "message": "ERROR",
  "data": "<Error Message>"
}
"""


def _browsing_prompt(jd_url: str) -> str:
    return f"""
Navigate to the provided URL: {jd_url} and extract the Job Description (JD).

Rules:
> If any popup / modal / dialog is found on the screen, strictly close it!
> If you don't find the webpage as JD of job you can throw an error.
> Extract only the required job description, strictly donot extract unwanted things like about the company etc.

{_static_prompt}
"""


async def agent_extract_jd(jd_url: str) -> str:
    logger.info(f"Starting agent_extract_jd() for {jd_url}")

    client = BrowserUse(api_key=settings.BROWSER_USE_API_KEY)

    logger.info(f"Navigating to {jd_url}")

    task = client.tasks.create_task(
        task=_browsing_prompt(jd_url),
        llm=GeminiModel.pro,  # Default model
    )

    # Run the blocking task completion in a separate thread
    result = await asyncio.to_thread(task.complete)

    extracted = result.output

    if not extracted:
        raise ValueError("Agent produced no output")

    try:
        # Simple cleanup to handle potential markdown code blocks
        cleaned_output = extracted.strip()
        if cleaned_output.startswith("```json"):
            cleaned_output = cleaned_output[7:]
        elif cleaned_output.startswith("```"):
            cleaned_output = cleaned_output[3:]

        if cleaned_output.endswith("```"):
            cleaned_output = cleaned_output[:-3]

        extracted_json = json.loads(cleaned_output.strip())
    except json.JSONDecodeError:
        logger.error(f"Failed to parse JSON from agent output: {extracted}")
        # Try to treat the whole output as data if JSON parsing fails, but respect the error protocol
        # If it failed to produce JSON, it might be a raw failure message or raw content.
        # Let's assume failure if not JSON, for safety, or wrap it.
        # But the prompt was strict.
        raise ValueError(f"Failed to parse agent output: {extracted}")

    jd_data = (
        extracted_json.get("data")
        if isinstance(extracted_json, dict)
        and extracted_json.get("message") == "SUCCESS"
        else None
    )

    if not jd_data:
        message = (
            extracted_json.get("data")
            if isinstance(extracted_json, dict)
            else "Unknown Error"
        )
        logger.error(f"Invalid response from agent: {message}")
        raise ValueError(message)

    logger.success(f"Extracted from {jd_url}")
    logger.info(f"Ending agent_extract_jd() for {jd_url}")

    return str(jd_data)

File: shared/ai/jd/extractJD.ts
import type { H3Event, EventHandlerRequest } from "h3";
import { chromium } from "playwright";
import { z } from "zod";
import { LLM } from "~/shared/integrations/llm";

interface JDConfig {
  url: string;
}

const BROWSING_PROMPT = `
      You are smart machine that understands and extracts contents from webpages.
      Below content should be a JD of a certain job.
      As I mentioned "Should", there can be posibility that you might a find a different webpage which might not be a JD page of example it can be a login page, a shopping card.

      Rules:

      > If you don't find the webpage as JD of job you can throw an error.
      > Extract only the required job description, strictly donot extract unwanted things like about the company etc.

      Strictly follow this json format output::

      Success:

      {
        message: 'SUCCESS',
        data: "<Extracted Content Of the JD>" // for e.g. This is a key role within the Global Deal Desk team, which owns deal strategy, quote-to-cash execution, and pricing operations. You’ll work closely with sales, product, legal, and operations teams to ensure that CPQ changes reflect business prior....
      }

      Error:

      {
        message: 'ERROR',
        data: "<Error Message>" // add error reason in this data property like -> UNKNOWN PAGE, DIFFICULT TO REACH, BLOCKERS CAME UP etc
      }
`;

const ACTION_PROMPT =
  "If any popup / modal / dialog is found on the screen, strictly close it!";

const EXTRACTION_SCHEMA = z.object({
  message: z.string(),
  data: z.string(),
});

const TIMEOUT = 60000;

export default async function (
  event: H3Event<EventHandlerRequest>,
  config: JDConfig
) {
  const { url } = config;
  const runtimeConfig = useRuntimeConfig(event);
  const { logger } = Logging.client;

  const CHROME_PATH = runtimeConfig.CHROME_PATH;
  const API_KEY = runtimeConfig.GEMINI_API_KEY;

  logger.info(`CHROME_PATH -> ${CHROME_PATH}`);
  logger.info(`API_KEY -> ${API_KEY}`);

  const isLocal = process.env.NODE_ENV === "development";

  const chromeExecutablePath = isLocal
    ? CHROME_PATH
    : chromium.executablePath();

  const { page, close } = await browsingAgent({
    apiKey: API_KEY,
    executablePath: chromeExecutablePath,
    model: LLM.model.gemini.FLASH_V25,
  });

  await page.goto(url, { timeout: TIMEOUT });
  logger.info(`naviagted to page -> ${url}`);

  await page.act({ action: ACTION_PROMPT });
  logger.info(`acted on page -> ${url}`);

  const { message, data } = await page.extract<typeof EXTRACTION_SCHEMA>({
    instruction: BROWSING_PROMPT,
    schema: EXTRACTION_SCHEMA,
    modelName: LLM.model.gemini.FLASH_V25,
  });

  logger.info(`message -> ${message}`);
  logger.info(`data -> ${data}`);

  await close();

  if (message === "ERROR")
    throw createError({ statusCode: 400, message: data });

  return data;
}
